modules/image-loading-optimization/module.php

```diff
@@ -0,0 +1,339 @@
+			'title' => esc_html__( 'Optimize Image Loading', 'elementor' ),
+			'tag' => esc_html__( 'Performance', 'elementor' ),
+			'description' => esc_html__( 'Applying `fetchpriority="high"` on LCP image and lazy loading on images below the fold.', 'elementor' ),
+			'release_status' => Experiments_Manager::RELEASE_STATUS_ALPHA,
+			'default' => Experiments_Manager::STATE_INACTIVE,
+		];
+	}
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		parent::__construct();
+
+		// Stop wp core logic.
+		add_action( 'init', [ $this, 'stop_core_fetchpriority_high_logic' ] );
+		add_filter( 'wp_lazy_loading_enabled', '__return_false' );
+
+		// Run optimization logic on header.
+		add_action( 'get_header', [ $this, 'set_buffer' ] );
+		// Ensure buffer is flushed (if any) before the content logic.
+		add_filter( 'the_content', [ $this, 'flush_header_buffer' ], 0 );
+
+		// Run optimization logic on content.
+		add_filter( 'wp_content_img_tag', [ $this, 'loading_optimization_image' ] );
+
+		// Run optimization logic on footer. Flushing of footer buffer will be handled by PHP script end default logic.
+		add_action( 'get_footer', [ $this, 'set_buffer' ] );
+	}
+
+	/**
+	 * Stop WordPress core fetchpriority logic by setting the wp_high_priority_element_flag flag to false.
+	 */
+	public function stop_core_fetchpriority_high_logic() {
+		// wp_high_priority_element_flag was only introduced in 6.3.0
+		if ( function_exists( 'wp_high_priority_element_flag' ) ) {
+			wp_high_priority_element_flag( false );
+		}
+	}
+
+	/**
+	 * Set buffer to handle header and footer content.
+	 */
+	public function set_buffer() {
+		ob_start( [ $this, 'handle_buffer_content' ] );
+	}
+
+	/**
+	 * This function ensure that buffer if any is flushed before the content is called.
+	 * This function behaves more like an action than a filter.
+	 *
+	 * @param string $content the content.
+	 * @return string We simply return the content from parameter.
+	 */
+	public function flush_header_buffer( $content ) {
+		$buffer_status = ob_get_status();
+		if ( ! empty( $buffer_status ) &&
+			1 === $buffer_status['type'] &&
+			get_class( $this ) . '::handle_buffer_content' === $buffer_status['name'] ) {
+			ob_end_flush();
+		}
+		return $content;
+	}
+
+	/**
+	 * Callback to handle image optimization logic on buffered content.
+	 *
+	 * @param string $buffer Buffered content.
+	 * @return string Content with optimized images.
+	 */
+	public function handle_buffer_content( $buffer ) {
+		return $this->filter_images( $buffer );
+	}
+
+	/**
+	 * Check for image in the content provided and apply optimization logic on them.
+	 *
+	 * @param string $content Content to be analyzed.
+	 * @return string Content with optimized images.
+	 */
+	private function filter_images( $content ) {
+		return preg_replace_callback(
+			'/<img\s[^>]+>/',
+			function ( $matches ) {
+				return $this->loading_optimization_image( $matches[0] );
+			},
+			$content
+		);
+	}
```

Comment:
Hi,

WordPress doesn't know which media file you are using, that is why you need to use regex.

Elementor, on the other hand, has several methods that handle images.

Have you considered hooking to the `elementor/image_size/get_attachment_image_html` [filter](https://github.com/elementor/elementor/blob/main/includes/controls/groups/image-size.php#L137)? This hooks accepts all the relevant data and generates the attachment HTML code.

Reply:
Hi @rami-elementor! Thanks for pointing this out. 

In order to accurately apply the correct optimizations to each image, we need to know what order they appear in the rendered HTML, which is why running `preg_replace` on all the rendered HTML after all the markup has been rendered is more reliable. In my testing, it doesn't look like all image are run through this filter. For example, the basic image gallery relies on [the gallery shortcode](https://github.com/elementor/elementor/blob/main/includes/widgets/image-gallery.php#L391), which doesn't seem like it will get passed through this filter.

Reply:
Ok, thank you for the clarification.